capsules/extra/src/sht4x.rs

```diff
@@ -0,0 +1,246 @@
+        crc ^= data[x];
+        for _i in 0..8 {
+            if (crc & 0x80) != 0 {
+                crc = crc << 1 ^ polynomial;
+            } else {
+                crc = crc << 1;
+            }
+        }
+    }
+    crc
+}
+
+pub struct SHT4x<'a, A: Alarm<'a>, I: i2c::I2CDevice> {
+    i2c: &'a I,
+    humidity_client: OptionalCell<&'a dyn kernel::hil::sensors::HumidityClient>,
+    temperature_client: OptionalCell<&'a dyn kernel::hil::sensors::TemperatureClient>,
+    state: Cell<State>,
+    buffer: TakeCell<'static, [u8]>,
+    read_temp: Cell<bool>,
+    read_hum: Cell<bool>,
+    alarm: &'a A,
+}
+
+impl<'a, A: Alarm<'a>, I: i2c::I2CDevice> SHT4x<'a, A, I> {
+    pub fn new(i2c: &'a I, buffer: &'static mut [u8], alarm: &'a A) -> SHT4x<'a, A, I> {
+        SHT4x {
+            i2c: i2c,
+            humidity_client: OptionalCell::empty(),
+            temperature_client: OptionalCell::empty(),
+            state: Cell::new(State::Idle),
+            buffer: TakeCell::new(buffer),
+            read_temp: Cell::new(false),
+            read_hum: Cell::new(false),
+            alarm: alarm,
+        }
+    }
+
+    fn read_humidity(&self) -> Result<(), ErrorCode> {
+        if self.read_hum.get() == true {
+            Err(ErrorCode::BUSY)
+        } else {
+            if self.state.get() == State::Idle {
+                self.read_hum.set(true);
+                self.read_temp_hum()
+            } else {
+                self.read_hum.set(true);
+                Ok(())
+            }
+        }
+    }
+
+    fn read_temperature(&self) -> Result<(), ErrorCode> {
+        if self.read_temp.get() == true {
+            Err(ErrorCode::BUSY)
+        } else {
+            if self.state.get() == State::Idle {
+                self.read_temp.set(true);
+                self.read_temp_hum()
+            } else {
+                self.read_temp.set(true);
+                Ok(())
+            }
+        }
+    }
+
+    fn read_temp_hum(&self) -> Result<(), ErrorCode> {
+        self.buffer.take().map_or_else(
+            || panic!("SHT4x No buffer available!"),
+            |buffer| {
+                self.state.set(State::Read);
+                self.i2c.enable();
+
+                // buffer[0] = ((Registers::MEASHIGHREP as u16) >> 8) as u8;
+                // buffer[1] = ((Registers::MEASHIGHREP as u16) & 0xff) as u8;
+                buffer[0] = Registers::MEASHIGHREP as u8;
+
+                // TODO verify errors
+                let _ = self.i2c.write(buffer, 1);
+
+                Ok(())
+            },
+        )
+    }
+}
+
+impl<'a, A: Alarm<'a>, I: i2c::I2CDevice> time::AlarmClient for SHT4x<'a, A, I> {
+    fn alarm(&self) {
+        let state = self.state.get();
+        match state {
+            State::Read => {
+                self.state.set(State::ReadData);
+                self.buffer.take().map_or_else(
+                    || panic!("SHT4x No buffer available!"),
+                    |buffer| {
```

Comment:
```suggestion
                self.buffer.take().map(
                    |buffer| {
```

No need to have those panic messages in there.